virtualaddr.go

```diff
@@ -0,0 +1,172 @@
+package tunnel
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"sync/atomic"
+
+	"github.com/koding/logging"
+)
+
+type listener struct {
+	net.Listener
+	*vaddrOptions
+
+	done int32
+
+	// ips keeps track of registered clients for ip-based routing;
+	// when last client is deleted from the ip routing map, we stop
+	// listening on connections
+	ips map[string]struct{}
+}
+
+func (vaddr *vaddrStorage) newListener(l net.Listener) *listener {
+	return &listener{
+		Listener:     l,
+		vaddrOptions: vaddr.vaddrOptions,
+		ips:          make(map[string]struct{}),
+	}
+}
+
+func (l *listener) serve() {
+	for {
+		conn, err := l.Accept()
+		if err != nil {
+			l.log.Error("failue listening on %q: %s", l.Addr(), err)
+			return
+		}
+
+		if atomic.LoadInt32(&l.done) != 0 {
+			l.log.Debug("stopped serving %q", l.Addr())
+			conn.Close()
+			return
+		}
+
+		l.connCh <- conn
+	}
+}
+
+func (l *listener) localAddr() string {
+	if addr, ok := l.Addr().(*net.TCPAddr); ok {
+		if addr.IP.Equal(net.IPv4zero) {
+			return fmt.Sprintf("127.0.0.1:%d", addr.Port)
+		}
+	}
+	return l.Addr().String()
+}
+
+func (l *listener) stop() {
+	if atomic.CompareAndSwapInt32(&l.done, 0, 1) {
+		// self-pipe to break out of serve loop
+		conn, err := net.DialTimeout("tcp", l.localAddr(), defaultTimeout)
```

Comment:
`l.Accept()` will block until new connection arrives, we issue a dummy connect, so `l.Accept` returns and we return from the loop [here](https://github.com/koding/tunnel/pull/12/files#diff-9ff35f94bcd60257e5e3009930af187eR56). We do this as we can't just simply close the listener, as it is managed by the user (user may add / remove and the again add the listener multiple times, we're going to use this property in tunnelserver).

It's also safe if `l` is already busy accepting other connections, and we return from the loop after a user connection and not this one - either `DialTimeout` will timeout, or the dummy connection gets closed before we request yamux session.

ref: https://cr.yp.to/docs/selfpipe.html